File: Android/java/doraemonkit/src/main/java/com/didichuxing/doraemonkit/kit/weaknetwork/SpeedLimitRequestBody.java
package com.didichuxing.doraemonkit.kit.weaknetwork;

import android.os.SystemClock;

import com.didichuxing.doraemonkit.okhttp_api.OkHttpWrap;

import java.io.IOException;

import okhttp3.MediaType;
import okhttp3.RequestBody;
import okio.Buffer;
import okio.BufferedSink;
import okio.ForwardingSink;
import okio.Sink;

/**
 * Created by xiandanin on 2019-05-09 18:35
 */
public class SpeedLimitRequestBody extends RequestBody {
    private long mSpeedByte;//b/s
    private RequestBody mRequestBody;
    private BufferedSink mBufferedSink;

    public SpeedLimitRequestBody(long speed, RequestBody source) {
        this.mRequestBody = source;
        this.mSpeedByte = speed * 1024;//转成字节
    }

    @Override
    public MediaType contentType() {
        return mRequestBody.contentType();
    }

    @Override
    public long contentLength() throws IOException {
        return mRequestBody.contentLength();
    }

    @Override
    public void writeTo(BufferedSink sink) throws IOException {
        if (mBufferedSink == null) {
            //mBufferedSink = Okio.buffer(sink(sink));
            //默认8K 精确到1K
            mBufferedSink = OkHttpWrap.INSTANCE.createByteCountBufferedSink(sink(sink), 1024L);
        }
        mRequestBody.writeTo(mBufferedSink);
        mBufferedSink.close();
    }

    private Sink sink(final BufferedSink sink) {
        return new ForwardingSink(sink) {
            private long cacheTotalBytesWritten;
            private long cacheStartTime;

            @Override
            public void write(Buffer source, long byteCount) throws IOException {
                if (cacheStartTime == 0) {
                    cacheStartTime = SystemClock.uptimeMillis();
                }

                super.write(source, byteCount);
                cacheTotalBytesWritten += byteCount;

                long endTime = SystemClock.uptimeMillis() - cacheStartTime;
                //如果在一秒内
                if (endTime <= 1000L) {
                    //大小就超出了限制
                    if (cacheTotalBytesWritten >= mSpeedByte) {
                        long sleep = 1000L - endTime;
                        SystemClock.sleep(sleep);

                        //重置计算
                        cacheStartTime = 0L;
                        cacheTotalBytesWritten = 0L;
                    }
                }
            }
        };
    }

}

File: Android/kotlin/doraemonkit/src/main/java/com/didichuxing/doraemonkit/kit/weaknetwork/SpeedLimitRequestBody.kt
package com.didichuxing.doraemonkit.kit.weaknetwork

import android.os.SystemClock
import okhttp3.MediaType
import okhttp3.RequestBody
import okio.Buffer
import okio.BufferedSink
import okio.ForwardingSink
import okio.Sink
import java.io.IOException


/**
 * Created by zhanys on 2020-06-09.
 */
class SpeedLimitRequestBody(speed: Long, private val mRequestBody: RequestBody) : RequestBody() {
    private val mSpeedByte: Long = speed * 1024  //b/s
    private var mBufferedSink: BufferedSink? = null

    override fun contentType(): MediaType? {
        return mRequestBody.contentType()
    }

    @Throws(IOException::class)
    override fun contentLength(): Long {
        return mRequestBody.contentLength()
    }

    @Throws(IOException::class)
    override fun writeTo(sink: BufferedSink) {
        if (mBufferedSink == null) {
            //mBufferedSink = Okio.buffer(sink(sink));
            //默认8K 精确到1K
            mBufferedSink = ByteCountBufferedSink(sink(sink), 1024L)
        }
        mBufferedSink?.apply {
            mRequestBody.writeTo(this)
            close()
        }
    }

    private fun sink(sink: BufferedSink): Sink {
        return object : ForwardingSink(sink) {
            private var cacheTotalBytesWritten: Long = 0
            private var cacheStartTime: Long = 0

            @Throws(IOException::class)
            override fun write(source: Buffer, byteCount: Long) {
                if (cacheStartTime == 0L) {
                    cacheStartTime = SystemClock.uptimeMillis()
                }
                super.write(source, byteCount)
                cacheTotalBytesWritten += byteCount
                val endTime = SystemClock.uptimeMillis() - cacheStartTime
                //如果在一秒内
                if (endTime <= 1000L) {
                    //大小就超出了限制
                    if (cacheTotalBytesWritten >= mSpeedByte) {
                        val sleep = 1000L - endTime
                        SystemClock.sleep(sleep)

                        //重置计算
                        cacheStartTime = 0L
                        cacheTotalBytesWritten = 0L
                    }
                }
            }
        }
    }
}
